### PR TITLE
Rework of external force/torque logic

### PR DESCRIPTION
**Pull request by ad-daniel**
**Related Issues**
This pull-request relates to issue #358

**Description**
As suggested in issue #358, if an external force is applied to a node without physics (typically a sensor which by default don't), and a node ancestor with physics exists, then it should be possible to apply the external force anyway.

The proposed solution goes up the chain until it finds a dynamic solid and does fix issue #358.

Currently, **it searches upwards, not sideways,** so if there's a Transform or a Group with multiple children (some with and some without physics), applying a force on a children wouldn't actually move the entire Group. (Like the stacked box/cylinder on the video)

The search can be extended sideways too, but in practice I don't really know if it's desirable because:
- If the Group/Transform is part of a Robot, then by the Physics rules (i.e _if any of the children nodes have physics, then the topmost node (the Robot) also needs to_), so searching upwards will always lead to a solid with physics to be found.
- If it's not part of a Robot, it means it's part of the environment and likely was _intended_ to have parts with physics and parts without, would it make sense for everything to be affected by the force? (assuming there's an ancestor with physics, otherwise can't be moved anyway)

Which begs the question if it's even necessary to search at all since we can go straight to the Robot node and test that one. 
The only reason for actually climbing the hierarchy that I can imagine is if it's desirable to apply the force as close to the cursor as possible. In which case an even more deep search should be done, in all directions, and shouldn't just involve node vicinity (with transforms it could very well not be physically near, despite being close node-wise). In short, reverting to the Robot node might be the more 'reasonable' option.

Basically, I don't know if the proposed solution is sufficient, or even overkill.

**Test world**
[force_logic.zip](https://github.com/cyberbotics/webots/files/5780206/force_logic.zip)

**Video**

https://user-images.githubusercontent.com/44834743/103865457-87609d00-50c4-11eb-82cb-9e3aa13fed37.mp4

Red have physics, green don't. The Stack behind is a Transform composed by a Box + Cylinder

https://user-images.githubusercontent.com/44834743/103865488-92b3c880-50c4-11eb-83e9-ffab10ea300b.mp4

### CODE REVIEW

**Comment by omichel** [comment]
I agree with you it makes sense to search upwards and not sideways.
Also we should not go straight up to the top Solid node (you mentioned the Robot node, but we should actually be more generic) as the force should be applied as close as possible to the cursor. This is especially useful if a robot or object has multiple joints (legs) and we want to apply the force/torque at the joint we click and not on the upper parts. So, I believe your approach so far seems good.
I will be happy to review this PR more in depth once its out of draft.

**Comment by ad-daniel** [comment]
I've reworded the text a few times, as I kept changing my mind, so might be quite different now. Basically what I'm wondering is if node-vicinity suffices, with transforms involved they could be physically nowhere near despite being close node-wise, the force ending up affecting something else regardless.

Thinking of a spider-like robot, I suppose then a downwards search should be done as well, picking among the two directions the one with less jumps?

**Comment by omichel** [comment]
I understand your point, but we should actually consider that a Solid node with no Physics (or a Group/Transform node) is somehow a part of the closest upper Solid node with Physics. The "real" object with which we want to physically interact with is actually a Solid node, including all its non-physics enabled children (which we consider to be part of the same mechanical body, e.g., in the same solid merger).

A downward search seems odd, as a Solid node without Physics **cannot** have children with Physics.

**Comment by ad-daniel** [comment]
I was thinking of something like a spider-leg where the segments are made out of LEDs like the e-pucks body mentioned above, but I see, unless the Physics in the SolidDevice are active it wouldn't be possible to chain them like that anyway if the end-point is supposed to interact with the ground.

In which case the proposed solution should suffice.

File: src/webots/gui/WbView3D.cpp [hunk 1]
```diff
@@ -1943,9 +1943,21 @@ void WbView3D::mouseMoveEvent(QMouseEvent *event) {
...
+      selectedSolid = dynamic_cast<WbSolid *>(node);
+      if (selectedSolid && selectedSolid->bodyMerger() != NULL)
+        break;
+
+      node = node->parentNode();
+      if (!node || node->level() < 1)  // abort the search at the top of this node chain
+        return;
+    }
+    if (!selectedSolid)
       return;
```
Comment (omichel) [inline src/webots/gui/WbView3D.cpp:1959]: I believe the logic is correct, however, the code could be slightly simplified:

```suggestion
    if (!node)
      return;
    WbSolid *selectedSolid;
    while (1) {
      selectedSolid = dynamic_cast<WbSolid *>(node);
      if (selectedSolid && selectedSolid->bodyMerger() != NULL)
        break;

      node = node->parentNode();
      if (!node || node->level() < 1)  // abort the search at the top of this node chain
        return;
    }
```

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/webots/gui/WbView3D.cpp (hunk 1 above)
Comment (ad-daniel) [reply src/webots/gui/WbView3D.cpp]: Indeed, thank you.

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you.
We would also need an entry in the change log about this bug fix.

**Review by omichel — APPROVED** [review]
Thank you!
